File: kotlin/step/backtracking/main_14889.kt
import kotlin.math.absoluteValue
import kotlin.math.min

var n = 0
var minValue = Int.MAX_VALUE
lateinit var arr: Array<IntArray>
lateinit var team1: BooleanArray

fun main() {
    n = readLine()!!.toInt()
    arr = Array<IntArray>(n) { readLine()!!.split(' ').map { it.toInt() }.toIntArray() }
    team1 = BooleanArray(n)
    dfs(0, 0)
    println(minValue)
}

fun dfs(depth: Int, index: Int) {
    if (depth == n / 2) {
        var sum = 0
        for (i in 0 until n) {
            for (j in i + 1 until n) {
                if (i != j) {
                    if (team1[i] && team1[j])
                        sum += (arr[i][j] + arr[j][i])
                    else if (!team1[i] && !team1[j])
                        sum -= (arr[i][j] + arr[j][i])
                }
            }
        }
        minValue = min(minValue, sum.absoluteValue)
    } else {
        for (i in index until n) {
            if (!team1[i]) {
                team1[i] = true
                dfs(depth + 1, i + 1)
                team1[i] = false
            }
        }
    }
}

File: cpp/step/backtracking/main_14889.cpp
#include <iostream>
#include <cmath>
using namespace std;

int n;
int** arr;
bool* team;
int minValue = 5000;

void dfs(const int depth, const int start) {
	if (depth == n / 2) {
		int sum = 0;
		for (int i = 0; i < n; ++i) {
			for (int j = 0; j < i; ++j) {
				if (i != j) {
					if (team[i] && team[j]) {
						sum += (arr[i][j] + arr[j][i]);
					} else if (!team[i] && !team[j]) {
						sum -= (arr[i][j] + arr[j][i]);
					}
				}
			}
		}
		minValue = min(minValue, abs(sum));
	} else {
		for (int i = start; i < n; ++i) {
			if (!team[i]) {
				team[i] = true;
				dfs(depth + 1, i + 1);
				team[i] = false;
			}
		}
	}
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);

	cin >> n;
	team = new bool[n] {false, };
	arr = new int* [n];
	for (int i = 0; i < n; ++i) {
		arr[i] = new int[n];
		for (int j = 0; j < n; ++j) {
			cin >>arr[i][j];
		}
	}
	dfs(0, 0);
	cout << minValue;
}
